### DISCUSSION TITLE
setNodes from useState x setNodes from useReactFlow hook - open & Pro examples

### DISCUSSION THREAD

**Discussion by felipelionmotta** [xyflow/xyflow#3167] [Q&A]
Across different examples, setNodes can come either from useState or from useReactFlow hook. On one single example, for instance on Pro Auto Layout, it comes from these two different sources.

When to use setNodes created in your own useState and when to use setNodes extracted from useReactFlow hook? Will they not interfere with each other?

**Comment by bcakmakoglu** [comment]
`setNodes` which is provided by `useReactFlow` sets the state directly, whereas `setNodes` provided by `useState` will update that state, which is then passed to React Flow which then updates the state.

Though it's mostly for convenience, say you want to call `setNodes` or `setEdges` from a component that is nested a couple levels deep into your app and you don't want to drill `setNodes` from `useState` all the way down to that component.

Instead you can just use the `useReactFlow` hook to access `setNodes` from there and have the same effect without drilling a prop n-levels deep.
